builtin/credential/ldap/backend.go

```diff
@@ -212,70 +251,89 @@ func getUserDN(cfg *ConfigEntry, c *ldap.Conn, bindDN string) (string, error) {
+		UserDN   string
+		Username string
+	}{
+		ldap.EscapeFilter(userDN),
+		ldap.EscapeFilter(username),
+	}
+
+	var renderedQuery bytes.Buffer
+	t.Execute(&renderedQuery, context)
+	cfg.Logger.Printf("[DEBUG] LDAP search with query: %s", renderedQuery.String())
+
+	// TODO - ok to proceed if renderedQuery length is 0?
 	result, err := c.Search(&ldap.SearchRequest{
-		BaseDN: userDN,
-		Scope:  0,        // base scope to fetch only the userDN
-		Filter: "(cn=*)", // bogus filter, required to fetch the CN from userDN
+		BaseDN: cfg.GroupDN,
+		Scope:  2, // subtree
+		Filter: renderedQuery.String(),
 		Attributes: []string{
-			"memberOf",
+			cfg.GroupAttr,
 		},
 	})
-	// this check remains in case something happens with the ldap query or connection
 	if err != nil {
-		return nil, fmt.Errorf("LDAP fetch of distinguishedName=%s failed: %v", userDN, err)
-	}
-	// if there are more than one entry, we consider the results irrelevant and ignore them
-	if len(result.Entries) == 1 {
-		for _, attr := range result.Entries[0].Attributes {
-			// Find the groups the user is member of from the 'memberOf' attribute extracting the CN
-			if attr.Name == "memberOf" {
-				for _, value := range attr.Values {
-					memberOfDN, err := ldap.ParseDN(value)
-					if err != nil || len(memberOfDN.RDNs) == 0 {
-						continue
-					}
-
-					for _, rdn := range memberOfDN.RDNs {
-						for _, rdnTypeAndValue := range rdn.Attributes {
-							if strings.EqualFold(rdnTypeAndValue.Type, "CN") {
-								ldapMap[rdnTypeAndValue.Value] = true
-							}
-						}
-					}
-				}
-			}
-		}
+		return nil, fmt.Errorf("LDAP search failed: %v", err)
 	}
 
-	// Find groups by searching in groupDN for any of the memberUid, member or uniqueMember attributes
-	// and retrieving the CN in the DN result
-	if cfg.GroupDN != "" {
-		result, err := c.Search(&ldap.SearchRequest{
-			BaseDN: cfg.GroupDN,
-			Scope:  2, // subtree
-			Filter: fmt.Sprintf("(|(memberUid=%s)(member=%s)(uniqueMember=%s))", ldap.EscapeFilter(username), ldap.EscapeFilter(userDN), ldap.EscapeFilter(userDN)),
-		})
-		if err != nil {
-			return nil, fmt.Errorf("LDAP search failed: %v", err)
+	for _, e := range result.Entries {
+		dn, err := ldap.ParseDN(e.DN)
+		if err != nil || len(dn.RDNs) == 0 {
+			continue
 		}
 
-		for _, e := range result.Entries {
-			dn, err := ldap.ParseDN(e.DN)
-			if err != nil || len(dn.RDNs) == 0 {
-				continue
-			}
-			for _, rdn := range dn.RDNs {
-				for _, rdnTypeAndValue := range rdn.Attributes {
-					if strings.EqualFold(rdnTypeAndValue.Type, "CN") {
-						ldapMap[rdnTypeAndValue.Value] = true
-					}
-				}
+		// Enumerate attributes of each result, parse out CN and add as group
+		values := e.GetAttributeValues(cfg.GroupAttr)
+		if len(values) > 0 {
+			for _, val := range values {
+				groupCN := getCN(val)
+				cfg.Logger.Printf("[DEBUG] Adding group %s", groupCN)
+				ldapMap[groupCN] = true
 			}
+		} else {
+			// If groupattr didn't resolve, use self
```

Comment:
The first case is enumerating `User` objects, and needs to follow `GroupAttr` to get to the groups.

The second case is enumerating `Group` objects, and just needs to parse them.